File: chat/src/main/java/org/thoughtcrime/securesms/util/BitmapUtil.java
package org.thoughtcrime.securesms.util;

import android.graphics.Bitmap;
import android.graphics.Bitmap.CompressFormat;

import androidx.annotation.Nullable;
import androidx.annotation.WorkerThread;

import com.difft.android.base.log.lumberjack.L;

import java.io.ByteArrayOutputStream;

public class BitmapUtil {

  private static final String TAG = "BitmapUtil";


  @WorkerThread
  public static Bitmap createScaledBitmap(Bitmap bitmap, int maxWidth, int maxHeight) {
    if (bitmap.getWidth() <= maxWidth && bitmap.getHeight() <= maxHeight) {
      return bitmap;
    }

    if (maxWidth <= 0 || maxHeight <= 0) {
      return bitmap;
    }

    int newWidth  = maxWidth;
    int newHeight = maxHeight;

    float widthRatio  = bitmap.getWidth()  / (float) maxWidth;
    float heightRatio = bitmap.getHeight() / (float) maxHeight;

    if (widthRatio > heightRatio) {
      newHeight = (int) (bitmap.getHeight() / widthRatio);
    } else {
      newWidth = (int) (bitmap.getWidth() / heightRatio);
    }

    return Bitmap.createScaledBitmap(bitmap, newWidth, newHeight, true);
  }


  public static @Nullable byte[] toByteArray(@Nullable Bitmap bitmap) {
    if (bitmap == null) return null;
    ByteArrayOutputStream stream = new ByteArrayOutputStream();
    bitmap.compress(CompressFormat.PNG, 100, stream);
    return stream.toByteArray();
  }

}

File: chat/src/main/java/com/difft/android/chat/util/BitmapUtil.kt
package com.difft.android.chat.util

import android.graphics.Bitmap
import androidx.annotation.WorkerThread
import java.io.ByteArrayOutputStream

object BitmapUtil {

    @JvmStatic
    @WorkerThread
    fun createScaledBitmap(bitmap: Bitmap, maxWidth: Int, maxHeight: Int): Bitmap {
        if (bitmap.width <= maxWidth && bitmap.height <= maxHeight) {
            return bitmap
        }

        if (maxWidth <= 0 || maxHeight <= 0) {
            return bitmap
        }

        var newWidth = maxWidth
        var newHeight = maxHeight

        val widthRatio = bitmap.width / maxWidth.toFloat()
        val heightRatio = bitmap.height / maxHeight.toFloat()

        if (widthRatio > heightRatio) {
            newHeight = (bitmap.height / widthRatio).toInt()
        } else {
            newWidth = (bitmap.width / heightRatio).toInt()
        }

        return Bitmap.createScaledBitmap(bitmap, newWidth, newHeight, true)
    }

    @JvmStatic
    fun toByteArray(bitmap: Bitmap?): ByteArray? {
        if (bitmap == null) return null
        val stream = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.PNG, 100, stream)
        return stream.toByteArray()
    }
}
